### PR TITLE
Make to_np_int32 generic for all dtypes

### PR DESCRIPTION
**Pull request by hyanwong**
Function to address #192

### CODE REVIEW

File: python/tskit/util.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+Module responsible for various utility functions used in other modules.
+"""
+
+import numpy as np
+
+
+def safe_np_int_cast(int_array, dtype, copy=True):
```
Comment (jeromekelleher) [inline python/tskit/util.py]: Probably want this to default to False I'd say - situations where we want to guarantee it's a copy are rare (and probably handled in the C code anyway).

File: python/tests/test_util.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
+# MIT License
+#
+# Copyright (c) 2018-2019 Tskit Developers
+# Copyright (c) 2017 University of Oxford
```
Comment (jeromekelleher) [inline python/tests/test_util.py]: Can delete this line, as this file is new.

File: python/tskit/util.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        # Since this is a new numpy array anyway, it's always a copy, so economize by
+        # setting copy=False
+        copy = False
+    if int_array.size == 0:
+        return int_array.astype(dtype, copy=copy)  # Allow empty arrays of any type
+    try:
+        return int_array.astype(dtype, casting='safe', copy=copy)
+    except TypeError:
+        bounds = np.iinfo(dtype)
+        if np.all(int_array >= bounds.min) and np.all(int_array <= bounds.max):
```
Comment (jeromekelleher) [inline python/tskit/util.py]: More idiomatic here to catch and throw the error, then continue the function as normal. Emphasises what the normal path is:
```python 
if np.any(int_array < bounds.min) or np.any(int_array > bounds.max):
    raise OverflowError("Cannot convert safely to {} type".format(dtype))
if int_array.dtype.kind == 'i' and np.dtype(dtype).kind == 'u':
# As before
```

File: python/tests/test_util.py (hunk 2 above)
Comment (hyanwong) [reply python/tests/test_util.py]: You mean just the 2017 line?

File: python/tskit/util.py (hunk 3 above)
Comment (hyanwong) [reply python/tskit/util.py]: Yes, nicer, I agree

File: python/tests/test_util.py (hunk 2 above)
Comment (jeromekelleher) [reply python/tests/test_util.py]: Yes, just the 2017 line.

Comment (jeromekelleher) [reply python/tests/test_util.py]: Maybe strictly this should be "Copyright (c) 2019 Tskit Developers" as well, since it's a new file.

**Comment by jeromekelleher** [comment]
Excellent, thanks @hyanwong. This is a very useful addition.

File: python/tests/test_util.py (hunk 2 above)
Comment (jeromekelleher) [reply python/tests/test_util.py]: Doesn't matter, ignore the last comments (didn't see it had been resolved).

**Comment by hyanwong** [comment]
@jeromekelleher - did you see my comment above. There's a weirdness at e.g. https://github.com/tskit-dev/tskit/blob/4d6fdf05b68bff19cabeb79340ef0842ce3dfdf7/python/tskit/trees.py#L3258 where 3 lines above the array is formed as uint32, then cast into int32. I'm not sure why this is, but it seems wrong. There are a few further places in the same file where this happens too.

**Comment by jeromekelleher** [comment]
I can't find the comment, but this is correct. It's two different arrays, right (sample_sets and sample_set_sizes)?

**Comment by hyanwong** [comment]
Sorry, you are right. I'm being thick.

**Pull request merged**